python/ray/data/_internal/plan.py

```diff
@@ -40,7 +40,7 @@
 INHERITABLE_REMOTE_ARGS = ["scheduling_strategy"]
 
 
-logger = logging.getLogger(__name__)
+logger = DatasetLogger(__name__)
```

Comment:
Meta comment: Instead of creating a custom logger implementation for what essentially amounts to adding an extra handler that dumps records to a file, couldn't we have some config that dumps Datasets-level logs to a file **in addition** to stdout? Then this would simplify to adding a logging handler if that config is set:
```suggestion
from ray._private.ray_constants import LOGGER_FORMAT, LOGGER_LEVEL

_logger = None

def _get_logger():
    global _logger
    
    if _logger is None:
        _logger = logging.getLogger(__name__)
        if ctx.enable_log_to_file:
            session_dir = global_node.get_session_dir_path()
            log_path = os.path.join(
                session_dir,
                "logs/ray-data.log"
            )
            file_log_handler = logging.FileHandler(log_path)
            file_log_formatter = logging.Formatter(fmt=LOGGER_FORMAT)
            file_log_handler.setFormatter(file_log_formatter)
            file_log_handler.setLevel(LOGGER_LEVEL)
            logger.addHandler(file_log_handler)
    return _logger
```

Note that this is formulated as a lazy logger initialization so we bypass the issue of Ray not yet being initialized.

Since the primary motivator for this feature is persistence of logs in the case of a job failure or crash, it seems like removing the existing (stdout) logging handlers is overkill, and potentially not desirable. Even if we wanted to support removing the stdout handler, that could easily be done with this same lazy logger initialization pattern, where we just remove all existing handlers before adding the file handler. In either case, we would obviate the need for the `DatasetLogger` proxy class.

Am I missing anything here?

Reply:
This largely makes sense to me, I can flesh this out more. One question here, where would be the appropriate place to call `_get_logger()`? in `ExecutionPlan.__init__()`?

Reply:
Also, we'd still need to keep the original initialization of `logger = logging.getLogger(__name__)` in order to have the logger created right? Calling `_get_logger()` is intended to be called in order to initialize the additional file-based handler?

Reply:
@scottjlee Just heard about the new requirement of wanting some things that are logged to a file to **not** be logged to stdout!

With that requirement in mind, a quick thought: if the plan is to use this `DatasetLogger` generically throughout the ray.data package, then the `DatasetLogger` abstraction seems like a nice helper, but I think that `log_to_stdout` should default to `True` to match the underlying Python logger’s behavior by default, otherwise we might be surprised that important log records are never showing up in stdout.

Reply:
Yeah agree to keep existing behavior, and make `log_to_stdout` default to `True`.

Reply:
Agreed, I enabled logging to stdout by default.